documentation/ir.md

```diff
@@ -0,0 +1,694 @@
+    nation_name
+    domestic_quantity = DEFAULT_TO($0.agg_0, 0)"]
+    E --> F --> G
+  end
+
+  subgraph C2["($0)"]
+    subgraph L1'["(H1')"]
+    direction LR
+        A'["Root"]
+    end
+
+    subgraph R1'["(H2')"]
+    direction LR
+        R'["Collection:
+        regions"]
+        S'["Where
+        name == 'EUROPE'"]
+        R' --> S'
+    end
+
+    subgraph L2'["(H3')"]
+    direction LR
+        E'["Sub-collection:
+        nations"]
+        F'["Calculate
+        nation_name: name"]
+        E' --> F'
+    end
+    subgraph L4["(H4)"]
+    direction LR
+      C["Sub-Collection:
+      customers"]
+    end
+    subgraph L5["(H5)"]
+    direction LR
+      D["Sub-Collection:
+      orders"]
+    end
+    subgraph L6["(H6)"]
+    direction LR
+      V1["Sub-Collection:
+      lines"]
+      V2["Where
+      $0.name == BACK(3).nation_name"]
+      V1 --> V2
+    end
+    subgraph C4["($0)"]
+    direction TB
+        subgraph H7["(H7)"]
+        Q1["Sub-Collection:
+      supplier"]
+        end
+        subgraph H8["(H8)"]
+        Q2["Sub-Collection:
+      nation"]
+        end
+      data2["Child Info:
+        join_keys: [(supplier_key, BACK(1).key)]"]
+    end
+    data1["Child Info:
+    join_keys: [(key, BACK(3).key), (BACK(1).key, BACK(4).key)]
+    agg_keys: [BACK(3).key, BACK(4).key]
+    aggs: {'agg_0': SUM(quantity)}"]
+  end
+
+    L1 --> R1
+    R1 --> L2
+    L1' --> R1'
+    R1' --> L2'
+    L2' -->L4
+    L2 -.->|Aggregation| C2
+    L4 --> L5
+    L5 --> L6
+    L6 -.->|Singular| C4
+    Q1 --> Q2
+```
+
+However, we can go one step further if we make an observation about this hybrid tree. Notice how now, we are computing the logic of `H1`, `H2`, and most of `H3` twice. This must happen if we intend to keep *every* nation, including the ones without any entries of `$0`. However, if we don't intend to keep nations where there are no records of `$0` to join onto, we can modify the original PyDough code as follows:
+
+```py
+selected_nations = regions.WHERE(
+  name == "EUROPE"
+).nations.CALCULATE(nation_name=name)
+domestic_lines = customers.orders.lines.WHERE(supplier.nation.name == nation_name)
+nation_domestic_purchase_info = selected_nations
+.WHERE(HAS(domestic_lines)).CALCULATE(
+    nation_name,
+    domestic_quantity=SUM(domestic_lines.quantity),
+)
+```
+
+With this modification, the original hybrid tree changes so the access to child `$0` of `H3` is now an `AGGREGATION_ONLY_MATCH` access, meaning we do an inner join instead of a left join. The de-correlation procedure will notice this and, after transforming child `$0` into its de-correlated form, realize it can do an optimization to delete the original `H1`, `H2` and prefix of `H3` since all of the data required from them is accessible inside `$0`. This is done via the hybrid "pull up" node, which specifies that the data comes from the specified child, instead of having existing data that gets joined with data from the child. This will look as follows:
```

Comment:
```suggestion
With this modification, the original hybrid tree is modified so that access to child `$0` of `H3` is now an `AGGREGATION_ONLY_MATCH` access, meaning we do an inner join instead of a left join. The de-correlation procedure will notice this and, after transforming child `$0` into its de-correlated form, recognize an opportunity to optimize by removing the original `H1`, `H2` and prefix of `H3` since all of the data required from them is accessible inside `$0`. This is done via the hybrid "pull up" node, which specifies that the data comes from the specified child, instead of having existing data that gets joined with data from the child. This will look as follows:
```